default_crypto.go

```diff
@@ -32,48 +32,58 @@ func (c DefaultCrypto) GenerateDH() (DHPair, error) {
 	var pubKey [32]byte
 	curve25519.ScalarBaseMult(&pubKey, &privKey)
 	return dhPair{
-		privateKey: privKey,
-		publicKey:  pubKey,
+		privateKey: privKey[:],
+		publicKey:  pubKey[:],
 	}, nil
 }
 
 // DH returns the output from the Diffie-Hellman calculation between
 // the private key from the DH key pair dhPair and the DH public key dbPub.
-func (c DefaultCrypto) DH(dhPair DHPair, dhPub Key) Key {
+func (c DefaultCrypto) DH(dhPair DHPair, dhPub Key) (Key, error) {
 	var (
 		dhOut   [32]byte
-		privKey [32]byte = dhPair.PrivateKey()
-		pubKey  [32]byte = dhPub
+		privKey [32]byte
+		pubKey  [32]byte
 	)
+	copy(privKey[:], dhPair.PrivateKey()[:32])
+	copy(pubKey[:], dhPub[:32])
```

Comment:
Shouldn't that 32 be a 33?

Ref: https://crypto.stackexchange.com/a/8916

Reply:
In this case it would be `32 bytes`, as this implementation uses `curve25519`, which compressed keys are only `32 bytes` https://tools.ietf.org/id/draft-josefsson-tls-curve25519-02.html#rfc.section.2.3 and  http://cr.yp.to/ecdh/curve25519-20060209.pdf . 

This caused a bit of a problem as we use instead `secp256k1` which compressed key has a length of `33 bytes`, but the type exposed was `32 bytes` (this implementation is not used at all in our production code, it's completely replaced by https://github.com/status-im/status-protocol-go/blob/master/crypto/ethereum_crypto.go ).

I have changed it to use a slice, which makes certain things easier and simplify the implementation, as it is converted to a slice when sending it over the wire because protobuf does not support fixed-size arrays.

But I could have also changed it to be `33 bytes`, which is the standard for elliptic keys, and just padded the array in this implementation (some libraries do that for better interop, appending the length for example).
Not sure what's best, what would you recommend?

Reply:
Ah.  OK.

> Not sure what's best, what would you recommend?

I would use a slice, as you have.  I would be concerned that the padding approach would hide errors, e.g., an out-of-bounds read doesn't panic because it reads into the padding.

Thanks for the explanation.

Reply:
thank you!

Reply:
One more thought: would it make sense to check that the length of `dhPub` is 32 and return an error if not (and similarly for `dhPair.PrivateKey()`)?

I realize that this implementation is just an example. But I think, in the long term, there would be real benefit to verifying that all of `status-protocol-go`'s tests pass using both `EthereumCrypto` and `DefaultCrypto`.

Reply:
sounds good, will do that, thanks